pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,394 @@
+			}
+			startIndex += batchSize
+		}
+	}
+}
+
+func (bfs *Service) fetchAndProcessBlocks(blockOids []oid.ID) error {
+	var wg sync.WaitGroup
+
+	for i, oidBlock := range blockOids {
+		wg.Add(1)
+		go func(oidBlock oid.ID, i int) {
+			defer wg.Done()
+
+			ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+			defer cancel()
+			rc, err := bfs.get(ctx, oidBlock.String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to fetch block %s", oidBlock.String()), zap.Error(err))
+				return
+			}
+
+			err = bfs.processBlock(rc, i)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to process block with OID %s", oidBlock.String()), zap.Error(err))
+			}
+		}(oidBlock, i)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (bfs *Service) fetchDataWithOid() error {
+	startIndex := bfs.chain.BlockHeight() / 2000
+	skip := bfs.chain.BlockHeight() % 2000
+
+	for {
+		select {
+		case <-bfs.quit:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.OidAttribute, fmt.Sprintf("%d", startIndex), object.MatchStringEqual)
+			prm.SetFilters(filters)
+
+			ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+			blockOidsObject, err := bfs.search(ctx, prm)
+			cancel()
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to find 'block_oids' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+			if len(blockOidsObject) == 0 {
+				bfs.log.Info(fmt.Sprintf("NeoFS block fetcher service: no %s object found with index %d, stopping.", bfs.cfg.OidAttribute, startIndex))
+				return errors.New("NeoFS block fetcher service: no 'block_oids' object found")
+			}
+
+			blockCtx, blockCancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+			defer blockCancel()
+			blockOidsData, err := bfs.get(blockCtx, blockOidsObject[0].String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to fetch %s object %d", bfs.cfg.OidAttribute, startIndex), zap.Error(err))
+				return err
+			}
+			blockOIDs, err := parseBlockOIDs(blockOidsData)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to parse %s object with index %d", bfs.cfg.OidAttribute, startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(blockOIDs) == 0 {
+				bfs.log.Info(fmt.Sprintf("NeoFS block fetcher service: no block found with index %d, stopping.", startIndex))
+				return errors.New("no block found")
+			}
+
+			blockOIDs = blockOIDs[skip:]
+
+			for len(blockOIDs) > 0 {
+				batch := blockOIDs
+				if len(blockOIDs) > bfs.cfg.BatchSize {
+					batch = blockOIDs[:bfs.cfg.BatchSize]
+					blockOIDs = blockOIDs[bfs.cfg.BatchSize:]
+				} else {
+					blockOIDs = nil
+				}
+
+				err = bfs.fetchAndProcessBlocks(batch)
+				if err != nil {
+					return err
+				}
+			}
```

Comment:
For this cycle you need to add `select` that will track `case <-bfs.quit:` and exit if an exit signal was sent to the server in the middle of subsequent batch downloading. `fetchData()` doesn't have this problem because it has batch-sized iteration.